File: problems/problems_3153/solution.ts
function sumDigitDifferences(nums: number[]): number {
	let length: number = 0;
	for (let num: number = nums[0]; num > 0; num = Math.floor(num / 10)) {
		length++;
	}
	const counter: Array<Array<number>> = new Array(length).fill(0).map(_ => new Array(10).fill(0));
	let ans: number = 0;
	for (let [i, num] of nums.entries()) {
		for (let j: number = 0; j < length; j++) {
			const d: number = num % 10;
			ans += i - counter[j][d];
			counter[j][d]++;
			num = Math.floor(num / 10);
		}
	}
	return ans;
};

export function Solve(inputJsonElement: string): any {
	const inputValues: string[] = inputJsonElement.split("\n");
	const nums: number[] = JSON.parse(inputValues[0]);
	return sumDigitDifferences(nums);
}

File: problems/problems_3153/solution.rs
use serde_json::{json, Value};

pub struct Solution;

impl Solution {
    pub fn sum_digit_differences(nums: Vec<i32>) -> i64 {
		let mut ans: i64 = 0;
		let mut length: usize = 0;
		let mut num: i32 = nums[0];
		while num > 0 {
			num = num / 10;
			length += 1;
		}
		let mut counter: Vec<Vec<i64>> = vec![vec![0; 10]; length];
		for i in 0..nums.len() {
			num = nums[i];
			for j in 0..length {
				let d: usize = num as usize % 10;
				ans += i as i64 - counter[j][d] as i64;
				counter[j][d] += 1;
				num = num / 10;
			}
		}
		ans
    }
}

#[cfg(feature = "solution_3153")]
pub fn solve(input_string: String) -> Value {
	let input_values: Vec<String> = input_string.split('\n').map(|x| x.to_string()).collect();
	let nums: Vec<i32> = serde_json::from_str(&input_values[0]).expect("Failed to parse input");
	json!(Solution::sum_digit_differences(nums))
}
